cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/BrokersInUseCheck.java

```diff
@@ -0,0 +1,104 @@
+import org.apache.kafka.common.TopicPartitionInfo;
+
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * Class which contains several utility function which check if broker scale down or role change can be done or not.
+ */
+public class BrokersInUseCheck {
+    /**
+     * Logger
+     */
+    private static final ReconciliationLogger LOGGER = ReconciliationLogger.create(BrokersInUseCheck.class.getName());
+
+    /**
+     * Checks if broker contains any partition replicas when scaling down
+     *
+     * @param reconciliation        Reconciliation marker
+     * @param vertx                 Vert.x instance
+     * @param secretOperator        Secret operator for working with Secrets
+     * @param adminClientProvider   Used to create the Admin client instance
+     *
+     * @return returns future set of node ids containing partition replicas based on the outcome of the check
+     */
+    public Future<Set<Integer>> brokersInUse(Reconciliation reconciliation, Vertx vertx, SecretOperator secretOperator, AdminClientProvider adminClientProvider) {
+        return ReconcilerUtils.clientSecrets(reconciliation, secretOperator)
+                .compose(adminClientSecrets -> {
+                    try {
+                        String bootstrapHostname = KafkaResources.bootstrapServiceName(reconciliation.name()) + "." + reconciliation.namespace() + ".svc:" + KafkaCluster.REPLICATION_PORT;
+                        LOGGER.debugCr(reconciliation, "Creating AdminClient for Kafka cluster in namespace {}", reconciliation.namespace());
+                        Admin kafkaAdmin = adminClientProvider.createAdminClient(bootstrapHostname, adminClientSecrets.resultAt(0), adminClientSecrets.resultAt(1), "cluster-operator");
+
+                        return topicNames(reconciliation, vertx, kafkaAdmin)
+                                .compose(names -> describeTopics(reconciliation, vertx, kafkaAdmin, names))
+                                .compose(topicDescriptions -> {
+                                    Set<Integer> brokersWithPartitionReplicas = new HashSet<>();
+
+                                    for (TopicDescription td : topicDescriptions.values()) {
+                                        for (TopicPartitionInfo pd : td.partitions()) {
+                                            for (org.apache.kafka.common.Node broker : pd.replicas()) {
+                                                brokersWithPartitionReplicas.add(broker.id());
+                                            }
+                                        }
+                                    }
+
+                                    kafkaAdmin.close();
+                                    return Future.succeededFuture(brokersWithPartitionReplicas);
+                                }).recover(error -> {
+                                    LOGGER.warnCr(reconciliation, "Failed to get list of brokers in use", error);
```

Comment:
isn't it failed because of retrieving the topics? not the brokers? Or I didn't get what you meant.

Reply:
It will recover after failure in listing the topics, describing them, or processing the descriptions. So the error says that it failed to get the list of brokers in use which is what the `brokerInUse` method does. Listing a single of the things that might fail - such as listing the topics for example might be misledading.